torch/csrc/jit/tensorexpr/registerizer.h

```diff
@@ -0,0 +1,354 @@
+    std::unordered_map<const Buf*, std::vector<AccessInfo*>> access_by_buf;
+    for (const auto& pair : candidates_) {
+      AccessInfo* info = pair.second;
+      // Now that analysis is complete we must normalize the costs by the parent
+      // Block we plan to insert the scalar var into.
+      info->store_cost = IRSimplifier::simplify(
+          new Div(info->store_cost, costByBlock_[info->parent]));
+
+      info->load_cost = IRSimplifier::simplify(
+          new Div(info->load_cost, costByBlock_[info->parent]));
+
+      access_by_buf[pair.second->buf].push_back(pair.second);
+    }
+
+    for (const auto& pair : access_by_buf) {
+      const Buf* buf = pair.first;
+      const std::vector<AccessInfo*>& accesses = pair.second;
+      for (auto* info : accesses) {
+        // Filter out low cost accesses.
+        if (info->store_cost->isConstant() &&
+            immediateAs<int>(info->store_cost) <= 1) {
+          continue;
+        }
+
+        bool any_overlap = false;
+        // We want to find possible overlaps from other accesses.
+        for (auto* i2 : accesses) {
+          if (info == i2) {
+            continue;
+          }
+
+          // All accesses to a buf must have the same dimensionality.
+          assert(info->indices.size() == i2->indices.size());
+
+          // They don't overlap if there is a guaranteed difference in any
+          // dimension.
+          bool overlap = true;
+          for (size_t i = 0; i < info->indices.size(); ++i) {
+            const Expr* diff = new Sub(info->indices[i], i2->indices[i]);
+            diff = IRSimplifier::simplify(diff);
+            if (diff->isConstant() && !immediateEquals(diff, 0)) {
+              overlap = false;
+              break;
+            }
+          }
+
+          if (overlap) {
+            any_overlap = true;
+            break;
+          }
+        }
+        if (!any_overlap) {
+          ret.push_back(info);
+        }
+      }
+    }
+    return ret;
+  }
+
+  std::unordered_map<SimplifierHashType, AccessInfo*> candidates_;
+  std::unordered_map<const Block*, const Expr*> costByBlock_;
+
+  const Expr* loopCost_;
+  const Block* enclosingBlock_;
+  HashProvider hasher_;
+};
+
+// Walks the IR an replaces a single Acccess with a local scalar Var.
+class TORCH_API RegisterizerReplacer : public IRMutator {
+ public:
+  RegisterizerReplacer(AccessInfo* i) : info_(i) {
+    var_ = new Var(info_->buf->name_hint() + "_", info_->buf->dtype());
+    var_wrapper_ = new Buf(var_, {}, info_->buf->dtype());
+    initializer_ = info_->stores.front();
+  }
+
+  const Expr* mutate(const Load* v) override {
+    if (v->buf() != info_->buf) {
+      return IRMutator::mutate(v);
+    }
+
+    // sanity check indices for the same buf must have the same dimensionality.
+    assert(v->indices().size() == info_->indices.size());
+    for (int i = 0; i < info_->indices.size(); ++i) {
+      if (!exprEquals(v->indices()[i], info_->indices[i])) {
```

Comment:
Minor: this might get quite expensive for a fairly large program. Maybe add a TODO to remind our-future-selves.

Reply:
Which part gets expensive, comparing indices?